src/engine/client/audio.cpp

```diff
@@ -0,0 +1,330 @@
+LPALISAUXILIARYEFFECTSLOT alIsAuxiliaryEffectSlot;
+LPALAUXILIARYEFFECTSLOTI alAuxiliaryEffectSloti;
+LPALAUXILIARYEFFECTSLOTIV alAuxiliaryEffectSlotiv;
+LPALAUXILIARYEFFECTSLOTF alAuxiliaryEffectSlotf;
+LPALAUXILIARYEFFECTSLOTFV alAuxiliaryEffectSlotfv;
+LPALGETAUXILIARYEFFECTSLOTI alGetAuxiliaryEffectSloti;
+LPALGETAUXILIARYEFFECTSLOTIV alGetAuxiliaryEffectSlotiv;
+LPALGETAUXILIARYEFFECTSLOTF alGetAuxiliaryEffectSlotf;
+LPALGETAUXILIARYEFFECTSLOTFV alGetAuxiliaryEffectSlotfv;
+
+AudioManager *AudioManager::instance_ = nullptr;
+
+AudioManager::AudioManager() {
+    memset(al_extensions_, false, MAX_AUDIO_EXT_SLOTS);
+
+    ALCdevice *device = alcOpenDevice(nullptr);
+    if(device == nullptr) {
+        Error("failed to open audio device, aborting audio initialisation!\n");
+    }
+
+    if(alcIsExtensionPresent(device, "ALC_EXT_EFX")) {
+        LogInfo("ALC_EXT_EFX detected\n");
+
+        alGenEffects = (LPALGENEFFECTS) alGetProcAddress("alGenEffects");
+        alDeleteEffects = (LPALDELETEEFFECTS) alGetProcAddress("alDeleteEffects");
+        alIsEffect = (LPALISEFFECT) alGetProcAddress("alIsEffect");
+        alEffecti = (LPALEFFECTI) alGetProcAddress("alEffecti");
+        alEffectf = (LPALEFFECTF) alGetProcAddress("alEffectf");
+
+        alGenAuxiliaryEffectSlots = (LPALGENAUXILIARYEFFECTSLOTS) alGetProcAddress("alGenAuxiliaryEffectSlots");
+        alDeleteAuxiliaryEffectSlots = (LPALDELETEAUXILIARYEFFECTSLOTS) alGetProcAddress(
+                "alDeleteAuxiliaryEffectSlots");
+        alIsAuxiliaryEffectSlot = (LPALISAUXILIARYEFFECTSLOT) alGetProcAddress("alIsAuxiliaryEffectSlot");
+        alAuxiliaryEffectSloti = (LPALAUXILIARYEFFECTSLOTI) alGetProcAddress("alAuxiliaryEffectSloti");
+
+        al_extensions_[AUDIO_EXT_EFX] = true;
+    }
+
+    int attr[]={
+            ALC_FREQUENCY, 44100, /* todo: tune this */
+            ALC_MAX_AUXILIARY_SENDS, 4,
+            0
+    };
+
+    ALCcontext *context = alcCreateContext(device, attr);
+    if(context == nullptr || !alcMakeContextCurrent(context)) {
+        Error("failed to create audio context, aborting audio initialisation!\n");
+    }
+
+    if(alIsExtensionPresent("AL_SOFT_buffer_samples")) {
+        LogInfo("AL_SOFT_buffer_samples detected\n");
+        al_extensions_[AUDIO_EXT_SOFT_BUFFER_SAMPLES] = true;
+    }
+
+    alDopplerFactor(4.f);
+    alDopplerVelocity(350.f);
+}
+
+AudioManager::~AudioManager() {
+    LogInfo("shutting down audio sub-system...\n");
+
+    //Audio_StopSounds();
+    //Audio_ClearSamples();
+
+    ALCcontext *context = alcGetCurrentContext();
+    if(context != nullptr) {
+        alcMakeContextCurrent(nullptr);
+        alcDestroyContext(context);
+
+        ALCdevice *device = alcGetContextsDevice(context);
+        if(device != nullptr) {
+            alcCloseDevice(device);
+        }
+    }
+}
+
+void AudioManager::CacheSample(const std::string &path, bool preserve) {
+    if(plIsEmptyString(path)) {
```

Comment:
This seems unnecessary.

Reply:
Yeah my bad, it's totally redundant here now.